### DISCUSSION TITLE
Problem with using S3 cache with Minio

### DISCUSSION THREAD

**Discussion by szpak** [renovatebot/renovate#19702] [Request Help] [answered]
Due to the problems with distributed cache support on the GitLab instance I use, I switched to the cache using a S3 bucket hosted by Minio. Unfortunately, despite settings RENOVATE_X_S3_ENDPOINT, the S3 client seems to try to connect to AWS:

```
 WARN: RepoCacheS3.read() - failure (repository=group/repo1)
       "err": {
         "name": "InvalidAccessKeyId",
         "$fault": "client",
         "$metadata": {"httpStatusCode": 403, "attempts": 1, "totalRetryDelay": 0},
         "Code": "InvalidAccessKeyId",
         "Key": "gitlab/group/repo1/cache.json",
         "BucketName": "renovate-cache-tmp",
         "Resource": "/renovate-cache-tmp/gitlab/group/repo1/cache.json",
         "RequestId": "1737C503134419C0",
         "HostId": "3c240dde-d2c6-4c13-1234-965a0e86c177",
         "message": "The Access Key Id you provided does not exist in our records.",
         "stack": "InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n    at throwDefaultError (/usr/src/app/node_modules/@aws-sdk/smithy-client/dist-cjs/default-error-handler.js:8:22)\n    at deserializeAws_restXmlGetObjectCommandError (/usr/src/app/node_modules/@aws-sdk/client-s3/dist-cjs/protocols/Aws_restXml.js:4356:51)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at /usr/src/app/node_modules/@aws-sdk/middleware-serde/dist-cjs/deserializerMiddleware.js:7:24\n    at /usr/src/app/node_modules/@aws-sdk/middleware-signing/dist-cjs/middleware.js:14:20\n    at StandardRetryStrategy.retry (/usr/src/app/node_modules/@aws-sdk/middleware-retry/dist-cjs/StandardRetryStrategy.js:51:46)\n    at /usr/src/app/node_modules/@aws-sdk/middleware-flexible-checksums/dist-cjs/flexibleChecksumsMiddleware.js:56:20\n    at /usr/src/app/node_modules/@aws-sdk/middleware-logger/dist-cjs/loggerMiddleware.js:5:22\n    at RepoCacheS3.read (/usr/src/app/node_modules/renovate/lib/util/cache/repository/impl/s3.ts:34:29)\n    at RepoCacheS3.load (/usr/src/app/node_modules/renovate/lib/util/cache/repository/impl/base.ts:38:27)\n    at initRepoCache (/usr/src/app/node_modules/renovate/lib/util/cache/repository/init.ts:26:5)\n    at initializeCaches (/usr/src/app/node_modules/renovate/lib/workers/repository/init/cache.ts:19:3)\n    at initRepo (/usr/src/app/node_modules/renovate/lib/workers/repository/init/index.ts:37:3)\n    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:47:14)\n    at attributes.repository (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:173:11)\n    at start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:158:7)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:18:22"
       }
```

There is no clear URL which is used for connection, so I cannot be sure. However,  the bucket in Minio is exposed as public.

My cache related configuration:
```
RENOVATE_REPOSITORY_CACHE=enabled
RENOVATE_REPOSITORY_CACHE_TYPE=s3://renovate-cache-tmp
RENOVATE_X_S3_ENDPOINT=http://my-host-with-minio:9000/
```

I tried to use `RENOVATE_X_S3_PATH_STYLE`, but without any change. As the client was complaining about missing AWS variables, I defined AWS_REGION, AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY to some arbitrary values (although, I'm, testing with the anonymously available bucket). Renovate 34.78.0 from the Docker image.

Do you know what could be wrong and how to force Renovate to use Minio bucket for caching?

**Comment by viceice** [comment, marked as answer]
anonymous s3 access isn't supported. you need to add valid credentials for minio

**Reply by szpak** [reply]
Thanks for a quick response. Generating the key helped to make Renovate connect to my Minio's S3 bucket.

However, S3 cache seems to be effectively ignored :-/.

Sample scenario. The original execution (no local file cache, an empty bucket):

```
DEBUG: RepoCacheS3.read() - No cached file found (repository=group/repo1)
DEBUG: RepoCacheBase.load() - expecting data of type 'string' received 'object' instead - skipping (repository=group/repo1)
...
DEBUG: Package cache statistics (repository=group/repo1)
       "get": {"count": 19, "avgMs": 5, "medianMs": 0, "maxMs": 23},
       "set": {"count": 22, "avgMs": 16, "medianMs": 5, "maxMs": 152}
...
       "totalRequests": 2265
```

Looks good. There is cache.json in the bucket (2,7KiB - surprisily small). The subsequent execution:

```
DEBUG: RepoCacheS3.read() - success (repository=group/repo1)
DEBUG: Repository cache is restored from revision 13 (repository=group/repo1)
...
DEBUG: Package cache statistics (repository=group/repo1)
       "get": {"count": 24, "avgMs": 37, "medianMs": 9, "maxMs": 129},
       "set": {"count": 0}
...
       "totalRequests": 20
```

Great. Much faster.

However, when I delete the local cache directory (to simulate execution on another GitLab runner without distributed cache configured) it seems that the cached information in S3 doesn't help much:

```
DEBUG: RepoCacheS3.read() - success (repository=group/repo1)
DEBUG: Repository cache is restored from revision 13 (repository=group/repo1)
...
DEBUG: Package cache statistics (repository=group/repo1)
       "get": {"count": 19, "avgMs": 5, "medianMs": 0, "maxMs": 24},
       "set": {"count": 22, "avgMs": 14, "medianMs": 3, "maxMs": 139}
...
       "totalRequests": 2260
```

Very close to the original run :-/. S3 seems to be ignored. It wasn't even touched on that run (no update in Minio).

Do you know why?

Maybe there is some ENV set to force local cache? But I don't see any:

```
RENOVATE_BASE_DIR=...
RENOVATE_PLATFORM=gitlab
RENOVATE_REPOSITORY_CACHE=enabled
RENOVATE_X_S3_ENDPOINT=http://my-host-with-minio:9000/
RENOVATE_REPOSITORY_CACHE_TYPE=s3://renovate-cache-tmp
AWS_REGION=ignored
AWS_SECRET_ACCESS_KEY=...
AWS_ACCESS_KEY_ID=...
```
(other seem to be not related to caching)

**Reply by rarkins** [reply]
S3 is a repo cache (Eg extracted dependencies) and not a lookup cache. Package caches are normally stored locally in disk by default, or you can specify a Redis URL to store them remotely

**Reply by szpak** [reply]
Thanks @rarkins for your explanation.

As in the `repositoryCacheType` property description, `local` stands side-by-side with `s3://`, I wonder, if there is a plan to enhance the s3 cache to be a full alternative for `local`, or it is rather use to share "repo cache" across different Renovate "instances"?

**Reply by rarkins** [reply]
S3 repo cache is intended for either multiple instances or stateless instances. Otherwise file based cache works simply and well

**Reply by szpak** [reply]
Ok. I created https://github.com/renovatebot/renovate/pull/19711/ to clarify it in the documentation.
